File: implementations/julia-square-matrix/TSPBenchmark/src/espprc.jl
using DataStructures

mutable struct Label{ T <: Real, VB <: Vector{Bool}, VT <: Vector }
    at::Int
    visits::VB
    ignore::Bool
    pred::Union{Nothing,Label}
    cost::Float64
    length::T
    q::Vector{Int}
    successors::VT
end

# label extension
function Label(d::MT, l::LT, at::Int) where MT <: TSPData where LT <: Label
    visits = copy(l.visits)
    visits[at] = true
    q = [ (((at-1) & 1 << (i-1)) > 0) ? r + 1 : r for (i,r) in enumerate(l.q) ]
    Label(at, visits, false, l,
          l.cost + getaux(d, l.at, at), l.length + dist(d, l.at, at), q, [])
end

# create empty label
function Label(d::MT, nresources::Int) where MT <: TSPData
    Label(1, zeros(Bool, d.n), false, nothing, 0.0, 0,
          [0 for i in 1:nresources], [])
end

function dominates(l1::T, l2::T) where T <: Label
    if l1.cost > l2.cost || l1.length > l2.length
        return false
    end
    for v ∈ l1.visits
        if v ∉ l2.visits
            return false
        end
    end
    for (r1, r2) in zip(l1.q, l2.q)
        if r1 > r2
            return false
        end
    end
    true
end

function marksuccessors(l::LT) where LT <: Label
    for s ∈ l.successors
        s.ignore = true
        marksuccessors(s)
    end
end

# update given list of labels with new label
# returns true if newlabel is added, false otherwise
function update!(labels::Vector{LT}, newlabel::LT) where LT <: Label
    i = 1
    while i ≤ length(labels)
        if dominates(labels[i], newlabel)
            return false
        end
        if dominates(newlabel, labels[i])
            marksuccessors(labels[i])
            if i < length(labels)
                labels[i] = labels[end]
            end
            pop!(labels)
        else
            i += 1
        end
    end
    # at this point we know that newlabel is not dominated, so we add it
    push!(labels, newlabel)
    true
end

struct ESPPRC{ MT <: TSPData, T <: Real }
    d::MT
    nresources::Int
    resourcecapacity::Int
    maxlen::T
end

function solve(e::ESPPRC{MT, T}) where MT <: TSPData where T <: Real
    # initial label
    l = Label(e.d, e.nresources)
    Q = Deque{Int}()
    inQ = zeros(Bool, e.d.n)
    push!(Q, 1)
    inQ[1] = true
    labels = [ Vector{Label{T, Vector{Bool}}}() for i in 1:e.d.n ]
    push!(labels[1], l)
    # main DP loop
    while ! isempty(Q)
        n = popfirst!(Q)
        inQ[n] = false
        for label ∈ labels[n]
            label.ignore && continue
            for succ ∈ 1:e.d.n
                (label.visits[succ] || succ == n) && continue
                # is it length-feasible?
                if label.length + dist(e.d, n, succ) + dist(e.d, succ, 1) >
                    e.maxlen
                    continue
                end
                # is it resource-feasible?
                rfeas = true
                for r ∈ 1:e.nresources
                    if ((succ-1) & (1 << (r-1))) > 0 &&
                        label.q[r] + 1 > e.resourcecapacity
                        rfeas = false
                        break
                    end
                end
                rfeas || continue
                # at this point we know the extension is feasible so we do it
                nl = Label(e.d, label, succ)
                added = update!(labels[succ], nl)
                added && push!(label.successors, nl)
                if added && (! inQ[succ]) && succ ≠ 1
                    push!(Q, succ)
                    inQ[succ] = true
                end
            end
            label.ignore = true
        end
    end
    Int(trunc(minimum(label.cost for label in labels[1])))
end

File: implementations/python-flat-matrix/espprc.py
import collections

class Label:
    resources = []
    # construct an initial label corresponding to an empty path
    def __init__(self):
        # vertex where this partial path is arriving at
        self.at = 0
        # already visited nodes
        self.visited = set()
        # has this label already been extended?
        self.ignore = False
        # predecessor; should point to a Label
        self.pred = self
        self.cost = 0
        self.length = 0
        # resource consumption
        self.q = [0 for x in self.resources]
        # successors for later recursive deletion when dominated
        self.successors = []

    # Returns True if self dominates other
    # pre-condition: self.at == other.at
    def dominates(self, other):
        if self.cost > other.cost or self.length > other.length:
            return False
        for v in self.visited:
            if not v in other.visited:
                return False
        for s, o in zip(self.q, other.q):
            if s > o:
                return False
        return True

    # Extend label to given vertex
    def extend(self, vertex, rc, d):
        nl = Label()
        nl.at = vertex
        nl.pred = self
        nl.visited = self.visited.union((vertex,))
        nl.cost = self.cost + rc(self.at, vertex)
        nl.length = self.length + d(self.at, vertex)
        # resource is consumed if i^th bit of vertex is 1
        nl.q = [ r + 1 if (vertex & (1 << i) > 0) else r
                 for i, r in enumerate(self.q) ]
        return nl

    def __repr__(self):
        seq = [self.at]
        tmp = self
        while tmp.pred != tmp:
            tmp = tmp.pred
            seq.append(tmp.at)
        seq.reverse()
        return 'Label at node ' + str(self.at) \
            + '\tcost = ' + str(self.cost) + '\tlength = ' + str(self.length) \
            + '\tq = ' + str(self.q) \
            + '\tsequence = ' + str(seq)

    def marksuccessors(self):
        for s in self.successors:
            s.ignore = True
            s.marksuccessors()

# * compare newlabel with collection of labels removing dominated labels as
#   necessary
# * adds newlabel to collection labels if it is not dominated
# * returns true if newlabel was added
def updatedominance(labels, newlabel):
    i = 0
    while i < len(labels):
        if labels[i].dominates(newlabel):
            return False
        elif newlabel.dominates(labels[i]):
            labels[i].marksuccessors()
            if i < len(labels) - 1:
                labels[i] = labels[-1]
            del labels[-1]
        else:
            i += 1
    # at this point no label has dominated newlabel so we add it
    labels.append(newlabel)
    return True
    
class ESPPRC:
    def __init__(self, n, d, rc, tour, nresources, resourcecapacity, maxlen):
        self.n = n
        self.d = d
        self.rc = rc
        self.nresources = nresources
        self.rescap = resourcecapacity
        self.maxlen = maxlen
        
    def solve(self):
        # initialise DP data: list of labels, queue, initial label, resources
        vertices = [ x for x in range(self.n) ]
        Q, Qset = collections.deque(), set([0])
        Q.append(0)
        labels = [ [] for x in vertices ]
        Label.resources = [ x for x in range(self.nresources) ]
        labels[0].append(Label())
        # step 3: run DP
        while len(Qset) > 0:
            n = Q.popleft()
            Qset.remove(n)
            for label in labels[n]:
                if label.ignore:
                    continue
                for succ in vertices:
                    if succ in label.visited or succ == n: continue
                    # succ is a candidate for extension; is it length-feasible?
                    if label.length + self.d(n, succ) + self.d(succ, 0) \
                       > self.maxlen:
                        continue
                    # is it resource-feasible?
                    rfeas = True
                    for i, r in enumerate(label.q):
                        if (succ & (1 << i) > 0) and r + 1 > self.rescap:
                            rfeas = False
                            break
                    if not rfeas: continue
                    # now we can actually extend the label!
                    nl = label.extend(succ, self.rc, self.d)
                    # Let's update dominance
                    added = updatedominance(labels[succ], nl)
                    if added:
                        label.successors.append(nl)
                    if added and not (succ in Qset) and succ != 0:
                        Qset.add(succ)
                        Q.append(succ)
                label.ignore = True
        # step 4: return distance of cheapest label as hash value.
        return min(label.cost for label in labels[0])
